liquidity_pool/src/liquidity.rs

```diff
@@ -234,75 +221,66 @@ pub trait LiquidityModule:
                 .direct(&initial_caller, asset_token_id, 0, &extra_asset_paid, &[]);
         }
 
-        // TODO: Instead of borrow_token_amount (i.e. 1:1 ratio), calculate how much collateral amount was repaid
-        debt_position.collateral_amount -= borrow_token_amount;
-        if debt_position.collateral_amount == 0 {
-            self.debt_positions().remove(debt_position_id);
+        borrow_position.lend_tokens.amount -= borrow_token_amount;
+        if borrow_position.lend_tokens.amount == 0 {
+            self.borrow_position(borrow_token_nonce).clear();
         } else {
-            let _ = self
-                .debt_positions()
-                .insert(debt_position_id.clone(), debt_position.clone());
+            self.borrow_position(borrow_token_nonce)
+                .set(&borrow_position);
         }
 
         self.send()
             .esdt_local_burn(borrow_token_id, borrow_token_nonce, borrow_token_amount);
 
-        // Same here, use calculated amount of repaid collateral instead of borrow_token_amount
-        Ok((
-            debt_position.collateral_identifier,
-            borrow_token_amount.clone(),
-            debt_position.timestamp,
-        )
-            .into())
+        self.send().direct(
+            &initial_caller,
+            borrow_position.lend_tokens.get_token_id_as_ref(),
+            borrow_position.lend_tokens.get_token_nonce(),
+            &borrow_token_amount,
+            &[],
+        );
+
+        Ok(())
     }
 
     #[only_owner]
     #[payable("*")]
     #[endpoint]
     fn liquidate(
         &self,
-        position_id: BoxedBytes,
+        borrow_position_nonce: u64,
+        initial_caller: Address,
         #[payment_token] token: TokenIdentifier,
         #[payment_amount] amount: Self::BigUint,
-    ) -> SCResult<LiquidateData<Self::BigUint>> {
+    ) -> SCResult<()> {
+        self.require_non_zero_address(&initial_caller)?;
         require!(amount > 0, "amount must be bigger then 0");
         require!(
             token == self.pool_asset().get(),
             "asset is not supported by this pool"
         );
 
-        let mut debt_position = self.debt_positions().get(&position_id).unwrap_or_default();
-
-        require!(
-            debt_position != DebtPosition::default(),
-            "invalid debt position id"
-        );
-        require!(
-            !debt_position.is_liquidated,
-            "position is already liquidated"
-        );
         require!(
-            debt_position.health_factor < self.health_factor_threshold().get(),
-            "the health factor is not low enough"
+            !self.borrow_position(borrow_position_nonce).is_empty(),
+            "Position is empty"
```

Comment:
Maybe change `empty` to `repaid or already liquidated`?

Reply:
done